src/TheSpace/HarbergerMarket.sol

```diff
@@ -69,92 +80,126 @@ abstract contract HarbergerMarket is Multicall {
-        require(property.ownerOf(tokenId_) == msg.sender, "Sender does not own property");
+    function setTaxConfig(ConfigOptions option, uint256 value) external onlyOwner {
+        taxConfig[option] = value;
+
+        emit Config(option, value);
+    }
+
+    // TODO: withraw community treasury
+
+    /**
+     * @dev Set the current price of an Harberger property with token id.
+     *
+     * Emits a {Price} event.
+     */
+    function setPrice(uint256 tokenId, uint256 price) external {
+        require(property.ownerOf(tokenId) == msg.sender, "Sender does not own property");
+        require(price != this.getPrice(tokenId), "Price is the same");
 
-        _setPrice(tokenId_, price_);
+        _setPrice(tokenId, price);
     }
 
     /**
      * @dev Returns the current price of an Harberger property with token id.
      */
-    function getPrice(uint256 tokenId_) external view returns (uint256 price) {
-        return taxRecord[tokenId_].price;
+    function getPrice(uint256 tokenId) external view returns (uint256 price) {
+        return taxRecord[tokenId].price;
+    }
+
+    /**
+     * @dev Returns the current owner of an Harberger property with token id.
+     */
+    function getOwner(uint256 tokenId) external view returns (address owner) {
+        return property.ownerOf(tokenId);
     }
 
     /**
      * @dev Purchase property with bid higher than current price. Clear tax for owner before transfer.
      * TODO: check security implications
      */
-    function bid(uint256 tokenId_, uint256 price_) external {
-        // TODO: mint token if not already exists
+    function bid(uint256 tokenId, uint256 price) external {
+        require(property.ownerOf(tokenId) != msg.sender, "Already owned");
 
-        require(price_ >= this.getPrice(tokenId_), "Price too low");
+        if (property.exists(tokenId)) {
+            uint256 askPrice = this.getPrice(tokenId);
+            require(price >= this.getPrice(tokenId), "Price too low");
 
-        // collect tax
-        bool success = this.collectTax(tokenId_);
+            // collect tax
+            bool success = this.collectTax(tokenId);
 
-        if (success) {
-            // successfully clear tax
-            currency.transferFrom(msg.sender, property.ownerOf(tokenId_), price_);
-            property.safeTransferByMarket(property.ownerOf(tokenId_), msg.sender, tokenId_);
-        } else {
-            // if failed, mint to current bidder
-            property.mint(msg.sender, tokenId_);
+            if (success) {
+                // successfully clear tax
+                currency.transferFrom(msg.sender, property.ownerOf(tokenId), askPrice);
+                property.safeTransferByMarket(property.ownerOf(tokenId), msg.sender, tokenId);
+
+                return;
+            }
         }
+
+        // if token does not exists yet, or token is defaulted
+        // mint token to current sender for free
+        property.mint(msg.sender, tokenId);
```

Comment:
`property.mint` has `onlyMarket` modifier, couldn't work here?

Reply:
`msg.sender` inside `mint` should be the address of the market right?

Reply:
got it